mava/components/jax/training/step.py

```diff
@@ -719,3 +951,283 @@ def config_class() -> Optional[Callable]:
+                trainer.store.epoch_update_fn,
+                (
+                    states.random_key,
+                    states.policy_params,
+                    states.critic_params,
+                    states.policy_opt_states,
+                    states.critic_opt_states,
+                    batch,
+                ),
+                (),
+                length=trainer.store.global_config.num_epochs,
+            )
+
+            # Set the metrics
+            metrics = jax.tree_map(jnp.mean, metrics)
+            metrics["norm_policy_params"] = optax.global_norm(states.policy_params)
+            metrics["norm_critic_params"] = optax.global_norm(states.critic_params)
+            metrics["observations_mean"] = jnp.mean(
+                utils.batch_concat(
+                    jax.tree_map(
+                        lambda x: jnp.abs(jnp.mean(x, axis=(0, 1))), observations
+                    ),
+                    num_batch_dims=0,
+                )
+            )
+            metrics["observations_std"] = jnp.mean(
+                utils.batch_concat(
+                    jax.tree_map(lambda x: jnp.std(x, axis=(0, 1)), observations),
+                    num_batch_dims=0,
+                )
+            )
+            metrics["rewards_mean"] = jax.tree_map(
+                lambda x: jnp.mean(jnp.abs(jnp.mean(x, axis=(0, 1)))), rewards
+            )
+            metrics["rewards_std"] = jax.tree_map(
+                lambda x: jnp.std(x, axis=(0, 1)), rewards
+            )
+
+            new_states = TrainingStateStatsSeparateNetworks(
+                policy_params=new_policy_params,
+                critic_params=new_critic_params,
+                policy_opt_states=new_policy_opt_states,
+                critic_opt_states=new_critic_opt_states,
+                random_key=new_key,
+                stats=stats
+            )
+            return new_states, metrics
+
+        def step(sample: reverb.ReplaySample) -> Tuple[Dict[str, jnp.ndarray]]:
+            """Step over the reverb sample and update the parameters / optimiser states.
+
+            Args:
+                sample: Reverb sample.
+
+            Returns:
+                Metrics from SGD step.
+            """
+
+            # Repeat training for the given number of epoch, taking a random
+            # permutation for every epoch.
+            networks = trainer.store.networks["networks"]
+            policy_params = {
+                net_key: networks[net_key].policy_params for net_key in networks.keys()
+            }
+            critic_params = {
+                net_key: networks[net_key].critic_params for net_key in networks.keys()
+            }
+            policy_opt_states = trainer.store.policy_opt_states
+            critic_opt_states = trainer.store.critic_opt_states
+            random_key, _ = jax.random.split(trainer.store.key)
+            stats = trainer.store.stats
+
+            states = TrainingStateStatsSeparateNetworks(
```

Comment:
Once we get this [PR](https://github.com/instadeepai/Mava/pull/741), you can remove support for shared networks. Therefore you can remove the SeperateNetworks tags.

Reply:
I have removed the support of separate Networks.

Reply:
Thanks :)